MDEV-15443 Fix wsrep XID read from rollback segment header

temeo opened the pull request:
The problem: If some transactions were done with wsrep_on=0, a
rollback segment header having the highest trx_id assigned might
store undefined wsrep XID. When reading the wsrep checkpoint
from InnodB, the undefined wsrep XID might be returned instead
of the highest valid one.

The fix is to check if the rollback segment header stores the
valid wsrep XID before assigning it to the output XID.

I submit this patch under BSD-new license.

dr-m commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -127,17 +127,13 @@ void trx_rseg_update_wsrep_checkpoint(const XID* xid)
 bool trx_rseg_read_wsrep_checkpoint(const trx_rsegf_t* rseg_header, XID& xid)
 {
-	xid.formatID = (int)mach_read_from_4(
+	int formatID = (int)mach_read_from_4(
```
I would prefer to avoid type casts, especially C-style ones.

temeo replied:
Agreed. However, I chose to follow the style used in the surrounding code. Not having a cast might emit a warning from compiler if -Wsign-conversion is enabled.

I can change this to static cast and fix the surrounding wsrep related code too, would that be ok?

dr-m replied:
Yes, please feel free to fix. I also just moved the old code around without removing the type conversion. I think that the `formatID` could be unsigned; in fact, I understood that Galera would only write it as 1. Maybe other values than 1 should be ignored?

dr-m commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -211,13 +207,11 @@ bool trx_rseg_read_wsrep_checkpoint(XID& xid)
+		if ((!found || id > max_id) &&
+		    trx_rseg_read_wsrep_checkpoint(rseg_header, xid)) {
+			max_id = id;
+			found = true;
```
If write-set replication is disabled after the database server has run for quite some time, then it could happen that a WSREP XID is already present in all rollback segment header pages. In that case, we would still seem to return a possibly older WSREP XID.

Let’s assume that rollback segment header page A contained the latest WSREP XID before `SET wsrep_on=0` was executed. Now, let us assume that after that, the latest commit is on rollback segment header page B, which also happened to contain some older WSREP XID.

It seems to me that we should choose the maximum WSREP XID in this case, instead of choosing the WSREP XID of the rollback segment header page with the latest commit.

For this, we would need something that can compare (sort) WSREP XIDs. Can you explain how it would work?

Similarly, we should choose the maximum binary log file and position, instead of choosing the one on the rollback segment header page that contains the latest commit. This one should be straightforward: `strcmp()` the file names and compare the binlog offsets as unsigned numbers.

@Thirunarayanan can implement these changes.

temeo replied:
Good catch.

XIDs can be sorted by sequence number. Function read_wsrep_xid_seqno() can be used to read the sequence number part. One thing which seems to be missing though is a byte order conversion for the seqno, now the seqno is always stored in host byte order. Byte order conversion should be implemented in couple of places:
* storage/innobase/trx/trx0rseg.cc:read_wsrep_xid_seqno(), 
* sql/wsrep_xid.cc:wsrep_xid_init()
* sql/wsrep_xid.cc:wsrep_xid_seqno()

There is also another thing which I noticed after submitting a patch: The UUID part of the XID may change if the server is taken out of the existing cluster and used to bootstrap a new cluster. In the original implementation this was not a problem as the XID was stored in the single place in sys header. However, now the XID resides in several rseg headers, and if they contain different UUIDs, it is impossible to decide which of those represents the current cluster. One solution for this problem would be keeping cached value of current cluster UUID and clear all wsrep XIDs in rseg headers if the newly written XID has different UUID. This should be safe since when the cluster UUID changes, the first event which triggers the trx_rseg_update_wsrep_checkpoint() call is always executed in isolation, i.e. no interleaving wsrep operations are allowed.

dr-m replied:
Thank you for the clarification. Maybe you could contribute a revised patch that deals with this issue.

If this turns out to be too complicated, we can still revert to writing the WSREP XID to the TRX_SYS page only. Commit performance would be slightly degraded for Galera, but not for single-master replication commits that store the binlog position.

temeo commented:
@dr-m Rebased and updated the pull request.

1) It appears that it is enough to scan rseg headers for XID with highest seqno. Rollback segment trx_id can be ignored (unless it is zero when the XID read can be skipped).
2) The revised patch deals with the case where XID UUID part changes by resetting existing XIDs in rseg headers (except for rseg 0) to all zeros.
3) Could not get rid of the casts. XID is defined in sql/handler.h and uses long for formatID, gtrid_length, bqual_length. mach_read_from_n return unsigned integers and with certain compiler options this could emit warnings. Replaced old style casts with static_cast<> .
3) Could not figure simple way to make byte conversion for seqno part in XID without creating a copy of XID before writing it into rseg header, so I chose not to implement in this patch.

dr-m commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -101,9 +111,15 @@ trx_rseg_update_wsrep_checkpoint(
+when it is guaranteed that there are no wsrep transactions committing.
+If the UUID part of the wsrep XID does not match to the UUIDs of XIDs already
+stored into rollback segments, all the remaining XIDs will be reset to empty
+value.
```
There is a typo in the function name in the comment. I think that it would be more relevant to mention the function `wsrep_set_SE_checkpoint()`. If that function is invoked at `wsrep_commit()` and other places, then this almost certainly looks like the wrong approach.

The idea of MDEV-15158 was to avoid writing the `TRX_SYS` page during normal operations, and instead have the InnoDB transaction commit write to exactly one of the rollback segment header pages, which is already being written anyway.

With this change, InnoDB would in the worst case update all rollback segment header pages on transaction commit. Granted, it only happens when the UUID part of the XID changes. But, shouldn’t we do this one-time adjustment at InnoDB startup, instead of doing it during normal operation (presumably, on the first transaction commit)?

dr-m commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -101,9 +111,15 @@ trx_rseg_update_wsrep_checkpoint(
 @param[in]	xid		WSREP XID */
 void trx_rseg_update_wsrep_checkpoint(const XID* xid)
 {
+	static byte current_xid_uuid[16] = {0, };
```
I would prefer a stack-allocated buffer that would be initialized by `memset()`.

temeo replied:
The idea of static variable is to maintain the value over calls to this function. As mentioned above, this function is called for TOI and membership view events only, so it is guaranteed that only single thread is accessing this function at the time (with suitable memory barriers).

dr-m commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -66,33 +73,36 @@ trx_rseg_update_wsrep_checkpoint(
 
 	mlog_write_ulint(TRX_RSEG_WSREP_XID_FORMAT + rseg_header,
-			 uint32_t(xid->formatID),
+			 static_cast<uint32_t>(xid->formatID),
```
IMO the C++ constructor-style casts are just fine.

temeo commented on storage/innobase/trx/trx0rseg.cc:
```diff
@@ -118,6 +134,27 @@ void trx_rseg_update_wsrep_checkpoint(const XID* xid)
+	}
+	read_wsrep_xid_uuid(xid, current_xid_uuid);
+
 	mtr.commit();
```
I don't see any problem in splitting this loop into separate mini-transactions, except that if the server crashes during this operation, the incomplete operation should somehow detected at server initialization. In that case, it should be enough to use first found wsrep XID as a authority and clear the rest of the XIDs whose UUID does not match to the first found. From the implementation it seems that the rseg 0 will always contain the UUID of the last seen group membership change event, so it should be used.

dr-m replied:
Given that the `wsrep_set_SE_checkpoint()` is a rare operation that is executed on DDL operations, we can keep the behaviour that is implemented in the patch. Deadlocks between InnoDB threads should not be an issue, because I think that this is the only operation after startup where multiple rollback segment header pages are concurrently locked.

dr-m commented:
I pushed a revised version of this to 10.3.6 in commit 67f6d40bd982f32d03490bb917916b4fa867a475. Thank you for the contribution!
